subsys/net/lib/sockets/sockets.c

```diff
@@ -158,11 +158,7 @@ int zsock_close_ctx(struct net_context *ctx)
 	 * as these are fail-free operations and we're closing
 	 * socket anyway.
 	 */
-	if (net_context_get_state(ctx) == NET_CONTEXT_LISTENING) {
```

Comment:
Commit message:

> > When closing the socket, zsock_close_ctx() would send a
> "send" net_context_accept() when the socket was in "LISTEN" state.

"send" is apparently a wrong word here.

> This would result in an error log message for "Identical
connection handler already found" because the connection
context was already created because the initial
zsock_listen_ctx() called net_context_accept() to set up
the socket at the beginning.

This isn't how it works. net_context_accept() just sets callbacks, in particular, called with NULL, just resets callback. If there're weird error messages [now], it may be a result of incorrect refactors lately.

There's a comment before this code block, explaining what it does. You can't just remove it, or you'll introduce rare, nondeterministic leak into the system.

Reply:
Since the intent is to clear the callbacks then why don't we just have a function/macro to clear the callbacks instead of a side affect call to an "accept" function? 

Instead, the existing code calls:

 **net_context_accept()->net_tcp_accept()->net_tcp_register()->net_conn_register() -->> ERROR("Identical connection handler found").** 

The levels of indirection to just clear callbacks are annoying but the end result is that the error message causes the function to return before it is even able to set the callbacks so the existing code isn't even doing what you say it is intended to do?

If your intent was to clear the callbacks then I suggest we create a very specific function to do this.

Reply:
Here is what I see you want to reset:

context->connect_cb
context->recv_cb

The rest of the callbacks will be cleared when the system is finished sending the FIN message.

Reply:
> Since the intent is to clear the callbacks then why don't we just have a function/macro to clear the callbacks instead of a side affect call to an "accept" function?

Well, net_context_accept() is not an "accept" function, it's exactly a function to set an accept callback. Just the same as net_context_recv(), which you didn't remove, isn't really a "receive" function, but a function to setup a receive callback.

I would need to review the current code, and history of its change since the initial socket code was written, to give more detailed suggestions on what would be the best way to do here.

Reply:
Ok, the suggested fix is to patch net_tcp_accept() to check for `cb` param being NULL, then set context->tcp->accept_cb like that.

Of course, better fix would be to call net_tcp_register() just once, and then let change the callback freely. But that's the way of cleaning up augean stables, because the right way to do it once is to call net_tcp_register() on listen(), not on accept(). But of course, it's too late to clean it up after ~2 years of being done wrong, so later in 1.14 cycle. So I just suggest to do a stop-gap fix per above.

Reply:
@pfalcon , does this meet your request?